CI deployment cleanup (pt. 2)

Holzhaus opened the pull request:
Corresponding website PR: mixxxdj/website#231

Holzhaus committed:
CI: Improve file names of package artifacts

Holzhaus committed:
tools/deploy.py: Add more file metadata for download manifest

Holzhaus committed:
CI: Update win64 artifact slug to match download page on website

Holzhaus committed:
CI: Use better paths for branch and tag build artifacts

Holzhaus committed:
tools/deploy.py: Add `--dirty` flag to git describe for local usage

Be-ing commented on tools/deploy.py:
```diff
@@ -178,8 +189,8 @@ def collect_manifest_data(job_data):
         # Filter out unrelated job outputs that don't start with "artifact-".
-        prefix, _, slug = output_name.partition("-")
-        if prefix != "artifact" or not slug:
+        prefix, _, artifact_slug = output_name.partition("-")
```
Does this have an issue with the `-`s in the `git describe` output?

Holzhaus replied:
That's unrelated, this parses the output names from here: https://github.com/Holzhaus/mixxx/blob/ci-deployment-cleanup/.github/workflows/build.yml#L122-L124

Be-ing commented on .github/workflows/build.yml:
```diff
@@ -339,10 +339,16 @@ jobs:
       run: >
+        if [[ "${GITHUB_REF}" =~ ^refs/tags/.* ]];
+        then
+          export DEPLOY_PATH='mixxx-{git_describe}/mixxx-{git_describe}-{package_slug}{ext}';
```
The directory path code is repeated below for the manifest path. How about factoring that out to an environment variable used by both GH Actions steps?

Holzhaus replied:
Not possible, the other code runs in a fresh VM.

Be-ing replied:
okay

Be-ing commented on .github/workflows/build.yml:
```diff
@@ -398,7 +398,7 @@ jobs:
       # Retrieve the metadata from the matrix job's outputs, merge them into a
       # single JSON document and then deploy to the server.
-      if: github.event_name == 'push' && env.SSH_PASSWORD != null
+      if: github.event_name == 'push'
```
What is the point of building the manifest without deploying it? Just for testing right now?

Holzhaus replied:
Doesn't hurt, that step takes far less than a second and it helps to catch errors and makes it easier to review the manifest output if someone makes a PR that changes the python script or the GItHub Actions workflow.

Holzhaus replied:
We could even remove the `if:` condition for the whole "prepare for deployment" step, but I don't really care.

Holzhaus replied:
I was wrong, it does hurt because it checks if the files in the manifest actually exist and fails otherwise. I'll remove the change.

Holzhaus commented:
Corresponding action run in upstream repo: https://github.com/mixxxdj/mixxx/actions/runs/753727557

Holzhaus commented:
The tag code path is tested here: https://github.com/Holzhaus/mixxx/runs/2362195248?check_suite_focus=true

(I only pushed the tag to my fork. To ensure that I won't accidently push it to upstream, I deleted it immediately from both my local repo and the remote after the workflow started)

Holzhaus commented:
Proposed Merge Strategy:
- If you reviewed both this PR and mixxxdj/website#231, then I'll remove the WIP commit from the website PR. Then we merge the website PR first and this PR immediately afterwards. The downloads will be displayed as unavailable on the website for a short amount of time (until the mixxxdj/mixxx merge commit has been built and a netlify rebuild has been triggered), but I think that is okay.

Be-ing commented:
The `git describe` output in the run testing the tagging was `2.3-beta-3684-gd7ab3f3dc3`. That doesn't seem right. On a commit that is tagged, `git describe` should simply output the tag name.

Be-ing commented:
We also need to upload the source DEB to the release PPA rather than the mixxxbetas PPA for tags.

Holzhaus committed:
CI: Make sure to fetch all tags

Although this should be unnecessary since with use `fetch-depth: 0` for
the checkout action, but unfortunately this does not seem to suffice
to get the currently pushed tag via `git describe`.

Possibly related: actions/checkout#290

Holzhaus commented:
> The `git describe` output in the run testing the tagging was `2.3-beta-3684-gd7ab3f3dc3`. That doesn't seem right. On a commit that is tagged, `git describe` should simply output the tag name.

Looks like there is a GitHub Actions bug that prevents us from using git describe when tags are pushed. :-(

I pushed the code to this branch, then pushed a tag for this.
On the branch push it works: https://github.com/Holzhaus/mixxx/actions/runs/756128435
On the workflow triggered by the tag push it doesn't: https://github.com/Holzhaus/mixxx/actions/runs/756126402

Even though the commit hash is the same.

I don't know how to continue here.

Be-ing commented:
Can we check if the branch push is on a tag?

Holzhaus commented:
> Can we check if the branch push is on a tag?

Maybe using:

    git describe --tags --candidates=0

Holzhaus commented on .github/workflows/build.yml:
```diff
@@ -148,6 +148,12 @@ jobs:
+      # Works around an issue where not the latest tag is not fetched when the
+      # workflow is triggered by a tag push event.
+      # Possibly related: actions/checkout#290
+      run: git fetch origin --force --tags
```
This seems to work: https://github.com/Holzhaus/mixxx/runs/2365251296?check_suite_focus=true#step:28:25

Be-ing commented:
Thank you for working on this.

Be-ing commented:
https://mixxx.org/download is working :+1:
